### ISSUE TITLE
`exactOptionalPropertyTypes` does not apply to constructor parameters

### ISSUE THREAD

**Issue by Sxxov** [microsoft/TypeScript#48952]
# Bug Report

### 🔎 Search Terms

exact, optional, property, types, constructor, parameters, arguments, undefined, ?

### 🕗 Version & Regression Information

4.4 & above (including the latest nightly, v4.7.0-dev.20220504)
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about optional class members
- I was unable to test this on prior versions because `exactOptionalPropertyTypes` was added in 4.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?useDefineForClassFields=true&target=99&exactOptionalPropertyTypes=true&ts=4.7.0-dev.20220504#code/MYGwhgzhAECC0G8BQBIADgVwEYgJbGjQCcB7NAfgC5oByLMImgbiVWBIDsIAXIjYbiSIAKVOmx4CAoQAVSFajQBmJEjQA0qAJSIAvkn1IA9EejsiRAKYCQAT2jcAFqQDuMMB2iWLQpB0sucMJaAHTEZNAAvNAYHAAmlkq4-nEsxqbJ5lY2tsLkOnEklhAcNNwOziSBHl4+ROpeAG6Wnk4kGADmjtBg0AnNIGTe0AC2uF3llgAeaNbluOWCfgFBodJEchHRsQlJKSxAA)

### 💻 Code

```ts
class A {
    public prop?: 'bar';

    constructor(
        public ctorProp?: 'foo',
    ) {}
}

// correctly throws an error
new A().prop = undefined;

// incorrectly(?) doesn't throw an error, even though a developer might expect it to
new A().ctorProp = undefined;
```

### 🙁 Actual behavior

Only the property which is defined on the class itself reports as an error, when `undefined` is assigned to them.

```ts
// correctly throws an error
new A().prop = undefined;

// incorrectly(?) doesn't throw an error, even though a developer might expect it to
new A().ctorProp = undefined;
```

The types on the class is incorrect as well.

```ts
// 'bar'
new A().prop;

// 'bar' | undefined
new A().ctorProp;
```

In the JS emitted, one can see that an assignment is made regardless if a parameter is passed into it.

```ts
class A {
    ctorProp;
    prop;
    constructor(ctorProp) {
        this.ctorProp = ctorProp;
    }
}
```

While typescript's error is consistent with the outputted code, this may not be the expected behaviour when `exactOptionalPropertyTypes` is enabled. Related: #47350.

### 🙂 Expected behavior

Properties defined in both styles should report an error, when `undefined` is assigned to them.

```ts
// correctly throws an error
new A().prop = undefined;

// correctly throws an error
new A().ctorProp = undefined;
```

The types on the class should have their behaviours matched as well.

```ts
// 'bar'
new A().prop;

// 'bar'
new A().ctorProp;
```

The JS emitted, while I'm not sure how viable it is, should only assign when a parameter is passed in.

```ts
class A {
    ctorProp;
    prop;
    constructor(ctorProp) {
        // if target is >= es2021
        this.ctorProp ??= ctorProp;

        // else
        // if (0 in arguments) this.ctorProp = ctorProp;
    }
}
```

**Comment by MartinJohns** [comment]
IMO this is a duplicate of #44548. Or at the very least it would first require #44548.

`exactOptionalPropertyTypes` does not apply for parameters. You can pass `undefined` as a value, and then you'd require additional logic to be emitted - something that TypeScript opted out quite a while ago.

**Comment by Sxxov** [comment]
Oops, could not find #44548 due to the setting being changed from `--strictOptionalProperties` to `--exactOptionalPropertyTypes`.

Though I don't think it's 100% a duplicate, I understand the viewpoint & don't believe this issue can be fixed without additional logic. It still is a big thorn in DX though imo when using the ctor params pattern.
